src/hotspot/share/code/dependencies.cpp

```diff
@@ -239,7 +239,7 @@ void Dependencies::assert_common_2(DepType dept,
       }
     }
   } else {
-    if (note_dep_seen(dept, x0) && note_dep_seen(dept, x1)) {
+    if (note_dep_seen(dept, x0)) {
```

Comment:
Why `x0` is special in all these cases?
Should we do `||` instead?
```
  if (note_dep_seen(dept, x0) || note_dep_seen(dept, x1)) {
```

Reply:
It's just a fast path approximate check which guards linear search over recorded dependencies. 
`note_dep_seen(dept, x0) == false` says that such a dependency hasn't been seen before, but `note_dep_seen(dept, x0) == true` requires a search over recorded dependencies to prove the argument is in the right position.

Another way to fix it is to perform all the queries beforehand:
```
bool dep_seen_x0 = note_dep_seen(dept, x0);
bool dep_seen_x1 = note_dep_seen(dept, x1);
if (dep_seen_x0 && dep_seen_x1) {
  ...
}
```
It should dramatically reduce the rate of false positives.

Reply:
Pushed the implementation which checks all the arguments. Let me know what you prefer.